rcl/test/rcl/test_logging_rosout.cpp

```diff
@@ -0,0 +1,189 @@
+    const char * namespace_ = "/ns";
+    this->node_ptr = new rcl_node_t;
+    *this->node_ptr = rcl_get_zero_initialized_node();
+    ret = rcl_node_init(this->node_ptr, name, namespace_, this->context_ptr, &node_options);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    // create rosout subscription
+    const rosidl_message_type_support_t * ts =
+      ROSIDL_GET_MSG_TYPE_SUPPORT(rcl_interfaces, msg, Log);
+    const char * topic = "rosout";
+    this->subscription_ptr = new rcl_subscription_t;
+    *this->subscription_ptr = rcl_get_zero_initialized_subscription();
+    rcl_subscription_options_t subscription_options = rcl_subscription_get_default_options();
+    ret = rcl_subscription_init(this->subscription_ptr, this->node_ptr, ts, topic, &subscription_options);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  }
+
+  void TearDown()
+  {
+    rcl_ret_t ret = rcl_subscription_fini(this->subscription_ptr, this->node_ptr);
+    delete this->subscription_ptr;
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ret = rcl_node_fini(this->node_ptr);
+    delete this->node_ptr;
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ret = rcl_shutdown(this->context_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ret = rcl_context_fini(this->context_ptr);
+    delete this->context_ptr;
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  }
+};
+
+void
+wait_for_subscription_to_be_ready(
+  rcl_subscription_t * subscription,
+  rcl_context_t * context,
+  size_t max_tries,
+  int64_t period_ms,
+  bool & success)
+{
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t ret =
+    rcl_wait_set_init(&wait_set, 1, 0, 0, 0, 0, 0, context, rcl_get_default_allocator());
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
+    rcl_ret_t ret = rcl_wait_set_fini(&wait_set);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+  size_t iteration = 0;
+  do {
+    ++iteration;
+    ret = rcl_wait_set_clear(&wait_set);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ret = rcl_wait_set_add_subscription(&wait_set, subscription, NULL);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ret = rcl_wait(&wait_set, RCL_MS_TO_NS(period_ms));
+    if (ret == RCL_RET_TIMEOUT) {
+      continue;
+    }
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    for (size_t i = 0; i < wait_set.size_of_subscriptions; ++i) {
+      if (wait_set.subscriptions[i] && wait_set.subscriptions[i] == subscription) {
+        success = true;
+        return;
+      }
+    }
+  } while (iteration < max_tries);
+  success = false;
+}
+
+/* Basic nominal test of having rosout logging globally enabled and locally enabled in a node
+ */
+TEST_F(CLASSNAME(TestLoggingRosoutFixture, RMW_IMPLEMENTATION), test_enable_global_rosout_enable_nodeoption) {
```

Comment:
nit: You can use a [parameterized test](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#value-parameterized-tests).

Reply:
> nit: You can use a [parameterized test](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#value-parameterized-tests).

I try it. But find below problem.
                                                                                                                                                                                                                                                 
After using value-parameterized-tests, it seems INSTANTIATE_TEST_CASE_P not work well with CLASSNAME (TestLoggingRosoutFixture, RMW_IMPLEMENTATION).
(use googletest inside ros2, repository: https://github.com/ament/googletest.git, branch: ros2.)

```
INSTANTIATE_TEST_CASE_P(InstantiationName, CLASSNAME(TestLoggingRosoutFixture, RMW_IMPLEMENTATION),
  testing::Values(
    param_t(0, nullptr, true, true),
    param_t(0, nullptr, false, false),
    param_t(s_argc, s_argv, true, false),
    param_t(s_argc, s_argv, false, false)),);
```

build failed as follows,

```
[ 20%] Building CXX object test/CMakeFiles/test_logging_rosout__rmw_opensplice_cpp.dir/rcl/test_logging_rosout.cpp.o
In file included from /root/workspace/install/src/gtest_vendor/include/gtest/gtest.h:65:0,
                 from /root/workspace/src/ros2/rcl/rcl/test/rcl/test_logging_rosout.cpp:15:
/root/workspace/src/ros2/rcl/rcl/test/rcl/test_logging_rosout.cpp:165:58: error: pasting ")" and "_EvalGenerator_" does not give a valid preprocessing token
   CLASSNAME (TestLoggingRosoutFixture, RMW_IMPLEMENTATION),
                                                          ^
/root/workspace/install/src/gtest_vendor/include/gtest/gtest-param-test.h:562:23: note: in definition of macro ‘INSTANTIATE_TEST_CASE_P’
       gtest_##prefix##test_case_name##_EvalGenerator_() { return generator; } \
                       ^~~~~~~~~~~~~~
/root/workspace/src/ros2/rcl/rcl/test/rcl/test_logging_rosout.cpp:165:58: error: pasting ")" and "_EvalGenerateName_" does not give a valid preprocessing token
   CLASSNAME (TestLoggingRosoutFixture, RMW_IMPLEMENTATION),
                                                          ^
/root/workspace/install/src/gtest_vendor/include/gtest/gtest-param-test.h:563:40: note: in definition of macro ‘INSTANTIATE_TEST_CASE_P’
   static ::std::string gtest_##prefix##test_case_name##_EvalGenerateName_( \
                                        ^~~~~~~~~~~~~~
/root/workspace/src/ros2/rcl/rcl/test/rcl/test_logging_rosout.cpp:165:58: error: pasting ")" and "_dummy_" does not give a valid preprocessing token
   CLASSNAME (TestLoggingRosoutFixture, RMW_IMPLEMENTATION),
                                                          ^
/root/workspace/install/src/gtest_vendor/include/gtest/gtest-param-test.h:568:30: note: in definition of macro ‘INSTANTIATE_TEST_CASE_P’
   static int gtest_##prefix##test_case_name##_dummy_ GTEST_ATTRIBUTE_UNUSED_ = \
                              ^~~~~~~~~~~~~~
...<snip>
```

If using the following code
```
INSTANTIATE_TEST_CASE_P(InstantiationName, TestLoggingRosoutFixture,
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^ <== Use fixed class name, it works well. (But if test case failed, it is not clear to show which implementation cause the error.)
  testing::Values(
    param_t(0, nullptr, true, true),
    param_t(0, nullptr, false, false),
    param_t(s_argc, s_argv, true, false),
    param_t(s_argc, s_argv, false, false)),);

```

Do you think it's acceptable?

Reply:
@ivanpauno 

I found another way to solve the above issue.  
Update "ament/googletest" as follows,
```
diff --git a/googletest/include/gtest/gtest-param-test.h b/googletest/include/gtest/gtest-param-test.h
index 8cc3dd7..c7dca22 100644
--- a/googletest/include/gtest/gtest-param-test.h
+++ b/googletest/include/gtest/gtest-param-test.h
@@ -557,23 +557,32 @@ internal::CartesianProductHolder10<Generator1, Generator2, Generator3,
 // alphanumeric characters or underscore. Because PrintToString adds quotes
 // to std::string and C strings, it won't work for these types.
 
+#define GTEST_TEST_CASE_NAME_EVAL_GENERATOR(prefix, test_case_name) \
+  gtest_##prefix##test_case_name##_EvalGenerator_
+
+#define GTEST_TEST_CASE_NAME_EVAL_GENERATOR_NAME(prefix, test_case_name) \
+  gtest_##prefix##test_case_name##_EvalGenerateName_
+
+#define GTEST_TEST_CASE_NAME_DUMMY(prefix, test_case_name) \
+  gtest_##prefix##test_case_name##_dummy_
+
 # define INSTANTIATE_TEST_CASE_P(prefix, test_case_name, generator, ...) \
   static ::testing::internal::ParamGenerator<test_case_name::ParamType> \
-      gtest_##prefix##test_case_name##_EvalGenerator_() { return generator; } \
-  static ::std::string gtest_##prefix##test_case_name##_EvalGenerateName_( \
+      GTEST_TEST_CASE_NAME_EVAL_GENERATOR(prefix, test_case_name)() { return generator; } \
+  static ::std::string GTEST_TEST_CASE_NAME_EVAL_GENERATOR_NAME(prefix, test_case_name) ( \
       const ::testing::TestParamInfo<test_case_name::ParamType>& info) { \
     return ::testing::internal::GetParamNameGen<test_case_name::ParamType> \
         (__VA_ARGS__)(info); \
   } \
-  static int gtest_##prefix##test_case_name##_dummy_ GTEST_ATTRIBUTE_UNUSED_ = \
+  static int GTEST_TEST_CASE_NAME_DUMMY(prefix, test_case_name) GTEST_ATTRIBUTE_UNUSED_ = \
       ::testing::UnitTest::GetInstance()->parameterized_test_registry(). \
           GetTestCasePatternHolder<test_case_name>(\
               #test_case_name, \
               ::testing::internal::CodeLocation(\
                   __FILE__, __LINE__))->AddTestCaseInstantiation(\
                       #prefix, \
-                      &gtest_##prefix##test_case_name##_EvalGenerator_, \
-                      &gtest_##prefix##test_case_name##_EvalGenerateName_, \
+                      &GTEST_TEST_CASE_NAME_EVAL_GENERATOR(prefix, test_case_name), \
+                      &GTEST_TEST_CASE_NAME_EVAL_GENERATOR_NAME(prefix, test_case_name), \
                       __FILE__, __LINE__)
 
 }  // namespace testing

```

the following code works well.
```
INSTANTIATE_TEST_CASE_P(InstantiationName, CLASSNAME(TestLoggingRosoutFixture, RMW_IMPLEMENTATION),
  testing::Values(
    param_t(0, nullptr, true, true),
    param_t(0, nullptr, false, false),
    param_t(s_argc, s_argv, true, false),
    param_t(s_argc, s_argv, false, false)),);
```

Do you think we need to update "https://github.com/ament/googletest.git" ?

Reply:
I wouldn't patch google test, as we use the upstream code.
Here is an example https://github.com/ros2/rcl/blob/master/rcl/test/rcl/test_get_actual_qos.cpp that use parameterized tests and works well in all the platforms.
It adds a little of boilerplate macros, but it works fine.

Reply:
> I wouldn't patch google test, as we use the upstream code.
> Here is an example https://github.com/ros2/rcl/blob/master/rcl/test/rcl/test_get_actual_qos.cpp that use parameterized tests and works well in all the platforms.
> It adds a little of boilerplate macros, but it works fine.

Thanks. I will update codes again refer to that example.